teps/0097-breakpoints-for-taskruns-and-pipelineruns.md

```diff
@@ -0,0 +1,317 @@
+breakpoint for failed step 0. Running this script would create `/tekton/run/0/out` and `/tekton/run/0/out.breakpointexit`.
+
+## PipelineRun Proposal
+
+Breakpoints would be enabled on PipelineRuns in a way that the logic of putting the breakpoints in place will be in the 
+hands of the TaskRun controller and the step entrypoint.
+
+Using the below spec we will look at how this will work with the following example.
+    
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: quarante-deux-deux
+spec:
+  debug:
+    breakpoint:
+        onFailure: true
+        beforeTask: ["task-build-archive"]
+        afterTask: ["deploy-frontend"]
+```
+
+### Breakpoint before or after a TaskRun
+
+Upon providing a before/after breakpoint for a particular Task in a PipelineRun, the TaskRun will inherit breakpoints as
+follows. 
+
+#### before TaskRun
+
+Adding the name of the Task to `beforeTask` spec will add a breakpoint before the first step of the TaskRun of the
+Task mentioned in the spec.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: quarante-deux-deux
+spec:
+  debug:
+    breakpoint:
+      beforeTask: ["task-build-archive"]
+```
+
+#### after TaskRun
+
+Adding the name of the Task to `afterTask` spec will add a breakpoint after the last step of the TaskRun of the 
+Task mentioned in the spec.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: quarante-deux-deux
+spec:
+  debug:
+    breakpoint:
+      afterTask: ["deploy-frontend"]
+```
+
+### Breakpoint on failure of a TaskRun
+
+When we provide `onFailure` as `true` in the `breakpoint` spec of the PipelineRun, this will be applied to all the children TaskRun under
+this PipelineRun and hence `onFailure` will be present in the breakpoints of all the children TaskRuns.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: quarante-deux-deux
+spec:
+  debug:
+    breakpoint:
+      onFailure: true
+```
+
+## User Stories
+
+### CLI integration
+The scenario below provides details on what the user should do to debug their failed task.
+
+#### Environment access
+1. Get Failed TaskRun name.
+2. Provide as an argument to the cli to start the TaskRun in debug mode. To get
+   ```
+   tkn taskrun debug failed-taskrun-1234 --on-failure --console-access
+   ```
+   In the above example we are recreating the TaskRun with the following patch.
+   ```yaml
+   metadata:
+     name: failed-taskrun-1234-debug
+   spec:
+    debug: 
+      breakpoint: 
+        onFailure: true
+   ```
+3. Once the TaskRun is created and a step fails, the TaskRun Pod will be in the Running state. Till that time the CLI will wait 
+   for the TaskRun to stop executing (due to failure) and go into the limbo state which would be leveraged for debugging.
+   The CLI will open a shell to the step container which would be a reimplementation of `kubectl exec -it`.
+
+## Alternatives
```

Comment:
It's nice to see alternatives but there's no pros and cons specified for them, so it's not clear why you picked the other option? I think the proposed option looks good, but it would be good to call out pros / cons for these, or if none is available mention it in a comment.

Reply:
Makes sense, I'll add these.

Reply:
Done ! :heavy_check_mark: